bazel/linux/test.bzl

```diff
@@ -3,12 +3,118 @@ load("//bazel/linux:qemu.bzl", "kernel_qemu_run")
 load("//bazel/utils:macro.bzl", "mconfig", "mcreate_rule")
 load("//bazel/utils:exec_test.bzl", "exec_test")
 load("//bazel/linux:bundles.bzl", "kunit_bundle")
+load("//bazel/linux:runner.bzl", "commands_and_runtime", "get_prepare_run_check")
+load("//bazel/linux:providers.bzl", "RuntimeBundleInfo", "RuntimeInfo")
+load("@bazel_skylib//lib:shell.bzl", "shell")
 
-def qemu_test(name, kernel_image, run, qemu_binary = None,
+def _test_runner(ctx):
+    script_begin = """#!/bin/bash
+set -eu -o pipefail
+GREEN="\033[32m"
+RED="\033[31m"
+RESET="\033[0m"
+failed_outputs=()
+junit=""
+tests=0
+failures=0
+run_test() {
+    local title="$1"
+    local script="$2"
+    shift 2
+    local args=("$@")
+    local failed=1
+    local output
+    local status
+    local start_time_ns
+    local stop_time_ns
+    local duration_ns
+    local duration_s
+    local error_msg=""
+
+    printf "%-73s" "${title}"
+    start_time_ns=$(($(date +%s%N)))
+
+    output=$("${script}" "${args[@]}" 2>&1) && failed=0
+
+    tests=$[tests + 1]
+    if [ $failed -eq 0 ]; then
+        status="${GREEN}PASSED${RESET}"
+    else
+        failures=$[failures + 1]
+        status="${RED}FAILED${RESET}"
+        failed_outputs+="==================== Test output for ${title}:\\n${output}\\n"
+        error_msg="<failure message=\\\"failed\\\"/>"
+    fi
+    stop_time_ns=$(($(date +%s%N)))
+    duration_ns=$(($stop_time_ns - $start_time_ns))
+    duration_s=$(echo "$duration_ns/10^9" | bc -l)
+    printf "%s in %.1fs\\n" "$status" "$duration_s"
+    junit+='
+        <testsuite name="'"${title}"'" tests="1" failures="'"${failed}"'">
```

Comment:
nit, and ok to not consider this in this PR, leave a TODO: but when we run a gunit test or py_test in the VM, those will output their own junit.xml file. We should check that environment variables are such they don't overwrite the file you generate, and ideally they'd be merged in the junit.xml file you are creating.

(I know, I know)

Reply:
Good point. I will have to verify that.